### ISSUE TITLE
[rfc] purpose of `seq_eqclass`

### ISSUE THREAD

**Issue by CohenCyril** [math-comp/math-comp#424]
Can someone (@ggonthier ?) explain the use of `seq_eqclass`?
https://github.com/math-comp/math-comp/blob/71e397e46b65c1c27a65471170d71f388c8a45f1/mathcomp/ssreflect/seq.v#L953
Is it a uniform inheritance condition problem? Is it linked with `{pred T}`.
Maybe the answer should be a comment in the code at `seq_eqclass` definition time.

**Comment by pi8027** [comment]
Without `seq_eqclass`, `pred_of_seq` has type `forall (T : eqType), seq (Equality.sort T) -> {pred T}`, thus is not uniform.

**Comment by ggonthier** [comment]
Indeed, I do confirm that this is now purely a uniform inheritance issue. As documented in detail in `ssrbool`, it is necessary for the `topred` projection of `predType`s to be a coercion. However `pred_of_seq` should not be a coercion from `seq T` to `pred T`, because that would make `seq T` an applicative predicate (see `ssrbool`). The `seq_eqclass` source used to prevent this, but the introduction of `{pred T}`, and the additional requirement that `topred` coercions be coercions to `{pred T}`, has made this redundant. So now `seq_eqclass` only serves to placate the uniform inheritance requirement, which is unfortunately still broken even in the current version of Coq.
